daemon/container_operations_unix.go

```diff
@@ -610,7 +610,25 @@ func (daemon *Daemon) getNetworkSandbox(container *container.Container) libnetwo
 // ConnectToNetwork connects a container to a network
 func (daemon *Daemon) ConnectToNetwork(container *container.Container, idOrName string) error {
 	if !container.Running {
-		return derr.ErrorCodeNotRunning.WithArgs(container.ID)
+		if container.HostConfig.NetworkMode.IsContainer() {
```

Comment:
Yes. we can avoid the code duplication by simply moving the `if !container.Running {` check to https://github.com/docker/docker/blob/master/daemon/container_operations_unix.go#L645.
This will avoid the code duplication. But the tricky part will be differentiate the common code-path for `connectToNetwork` both from `docker run` command and `docker network connect`. We may have to check for a better container state than `container.Running` because it might be false when called via `docker run` & hence might cause issues.